Tab correction

fangly opened the pull request:
Hi Jim,

Please have a look at this PR which adds proper handling for tab character in source code. As you know, the R parser expands tab characters into a variable number of space characters (for 8-character wide indentation). This PR fixes the column numbers in the parsed_content table.

This is a change to a central piece of code, hence the PR. I have added various tests for this and the complete test suite passes successfully.

Cheers,

Florent

jimhester commented on tests/testthat/test-get_source_expressions.R:
```diff
@@ -0,0 +1,42 @@
+  f <- tempfile()
+
+  writeLines("1\n\t", f)
+  expect_error(lintr:::get_source_expressions(f), NA, label="empty parsed_content line")
```
We can remove the `lintr:::`, also it feels like it might be more readable if the redundant code was to a `compare_parsed_content` function

```r
compare_parsed_content <- function(content, code) {
  f <- tempfile()
  on.exit(unlink(f))
  writeLines(content, f)
  pc <- get_source_expressions(f)[["expressions"]][[1L]][["parsed_content"]]
  force(code)
}

fangly commented:
Hi Jim,

I have addressed your comments, fixed a bug I discovered (for tokens that span multiple lines) and worked on making the code faster.

Without tab indentation correction (code commented out):
```R
> setwd("tests/testthat")
> f1 <- "exclusions-test"
> f2 <- "exclusions-test-tabs"  # similar to previous file but with some tabs
Unit: milliseconds
                       expr      min       lq     mean   median       uq       max neval
 get_source_expressions(f1) 14.33412 15.30424 17.80299 16.23822 18.81035 173.48826  1000
 get_source_expressions(f2) 14.36000 15.21210 17.21882 16.03942 18.33064  58.05358  1000
```
With tab indentation correction:
```R
> microbenchmark::microbenchmark(get_source_expressions(f1), get_source_expressions(f2), unit = "ms", times = 1000L)
Unit: milliseconds
                       expr      min       lq     mean   median       uq      max neval
 get_source_expressions(f1) 14.53448 15.40911 17.50975 16.25920 18.23091 145.9753  1000
 get_source_expressions(f2) 15.05346 15.84531 18.06934 16.72239 19.05757  80.4700  1000
```

As you see, when getting the source expressions from files without tabs, the tab indentation correction code has no visible performance impact, and on file that have tabs, there is a small execution time (~3% in this case). When you run a full lint process (`lint()` with some linters), this speed decrease will become even less visible.

Cheers,

Florent
